// @ionic/cli-utils/lib/project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = require("path");
const chalk_1 = require("chalk");
const config_1 = require("./config");
const errors_1 = require("./errors");
const fs_1 = require("@ionic/cli-framework/utils/fs");
const npm_1 = require("@ionic/cli-framework/utils/npm");
const format_1 = require("./utils/format");
exports.PROJECT_FILE = 'ionic.config.json';
exports.PROJECT_FILE_LEGACY = 'ionic.project';
exports.PROJECT_TYPES = ['ionic-angular', 'ionic1', 'custom'];
class Project extends config_1.BaseConfig {
    loadAppId() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const p = yield this.load();
            if (!p.app_id) {
                throw new errors_1.FatalException(`Your project file (${chalk_1.default.bold(format_1.prettyPath(this.filePath))}) does not contain '${chalk_1.default.bold('app_id')}'. `
                    + `Run ${chalk_1.default.green('ionic link')}.`);
            }
            return p.app_id;
        });
    }
    loadPackageJson() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.packageJsonFile) {
                const packageJsonPath = path.resolve(this.directory, 'package.json');
                try {
                    this.packageJsonFile = yield npm_1.readPackageJsonFile(packageJsonPath);
                }
                catch (e) {
                    if (e === fs_1.ERROR_FILE_INVALID_JSON) {
                        throw new errors_1.FatalException(`Could not parse ${chalk_1.default.bold('package.json')}. Is it a valid JSON file?`);
                    }
                    else if (e === npm_1.ERROR_INVALID_PACKAGE_JSON) {
                        throw new errors_1.FatalException(`The ${chalk_1.default.bold('package.json')} file seems malformed.`);
                    }
                    throw e; // Probably file not found
                }
            }
            return this.packageJsonFile;
        });
    }
    loadBowerJson() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.bowerJsonFile) {
                const bowerJsonPath = path.resolve(this.directory, 'bower.json');
                try {
                    this.bowerJsonFile = yield npm_1.readBowerJsonFile(bowerJsonPath);
                }
                catch (e) {
                    if (e === fs_1.ERROR_FILE_INVALID_JSON) {
                        throw new errors_1.FatalException(`Could not parse ${chalk_1.default.bold('bower.json')}. Is it a valid JSON file?`);
                    }
                    else if (e === npm_1.ERROR_INVALID_BOWER_JSON) {
                        throw new errors_1.FatalException(`The ${chalk_1.default.bold('bower.json')} file seems malformed.`);
                    }
                    throw e; // Probably file not found
                }
            }
            return this.bowerJsonFile;
        });
    }
    provideDefaults(o) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const cloneDeep = yield Promise.resolve().then(function () { return require('lodash/cloneDeep'); });
            const results = cloneDeep(o);
            if (!results.name) {
                results.name = '';
            }
            if (!results.app_id) {
                results.app_id = '';
            }
            if (!results.integrations) {
                results.integrations = {};
            }
            if (!results.type) {
                results.type = yield this.determineType();
            }
            if (results.gulpFile) {
                results.integrations.gulp = { file: results.gulpFile };
            }
            if (results.gulp && typeof results.gulp.enabled === 'undefined') {
                results.gulp.enabled = true;
            }
            delete results.gulpFile;
            delete results.projectTypeId;
            delete results.typescript;
            delete results.v2;
            return results;
        });
    }
    getSourceDir() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const project = yield this.load();
            if (project.documentRoot) {
                return path.resolve(this.directory, project.documentRoot);
            }
            if (project.type === 'ionic1') {
                return path.resolve(this.directory, 'www');
            }
            return path.resolve(this.directory, 'src');
        });
    }
    is(j) {
        return j && typeof j.name === 'string' && typeof j.app_id === 'string';
    }
    formatType(type) {
        if (type === 'ionic-angular') {
            return 'Ionic Angular';
        }
        else if (type === 'ionic1') {
            return 'Ionic 1';
        }
        return type;
    }
    determineType() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                const packageJson = yield this.loadPackageJson();
                if (packageJson.dependencies && typeof packageJson.dependencies['ionic-angular'] === 'string') {
                    return 'ionic-angular';
                }
            }
            catch (e) {
                if (e.fatal) {
                    throw e;
                }
            }
            try {
                const bowerJson = yield this.loadBowerJson();
                if ((bowerJson.dependencies && typeof bowerJson.dependencies['ionic'] === 'string') || (bowerJson.devDependencies && typeof bowerJson.devDependencies['ionic'] === 'string')) {
                    return 'ionic1';
                }
            }
            catch (e) {
                if (e.fatal) {
                    throw e;
                }
            }
            throw new errors_1.FatalException(`Could not determine project type (project config: ${chalk_1.default.bold(format_1.prettyPath(this.filePath))}).\n` +
                `For ${this.formatType('ionic-angular')} projects, make sure ${chalk_1.default.green('ionic-angular')} is listed as a dependency in ${chalk_1.default.bold('package.json')}.\n` +
                `For ${this.formatType('ionic1')} projects, make sure ${chalk_1.default.green('ionic')} is listed as a dependency in ${chalk_1.default.bold('bower.json')}.\n\n` +
                `Alternatively, set ${chalk_1.default.bold('type')} attribute in ${chalk_1.default.bold('ionic.config.json')} to one of: ${exports.PROJECT_TYPES.map(v => chalk_1.default.green(v)).join(', ')}.\n\n` +
                `If the Ionic CLI does not know what type of project this is, ${chalk_1.default.green('ionic build')}, ${chalk_1.default.green('ionic serve')}, and other commands may not work. You can use the ${chalk_1.default.green('custom')} project type if that's okay.\n`);
        });
    }
}
exports.Project = Project;
