// ionic/dist/commands/cordova/build.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk_1 = require("chalk");
const command_1 = require("@ionic/cli-utils/lib/command");
const utils_1 = require("@ionic/cli-utils/lib/cordova/utils");
const app_scripts_1 = require("@ionic/cli-utils/lib/ionic-angular/app-scripts");
const base_1 = require("./base");
let BuildCommand = class BuildCommand extends base_1.CordovaCommand {
    preRun(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.preRunChecks();
            if (!inputs[0]) {
                const platform = yield this.env.prompt({
                    type: 'input',
                    name: 'platform',
                    message: `What platform would you like to build (${['android', 'ios'].map(v => chalk_1.default.green(v)).join(', ')}):`
                });
                inputs[0] = platform.trim();
            }
            yield this.checkForPlatformInstallation(inputs[0]);
        });
    }
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (options.build) {
                const { build } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/commands/build'); });
                yield build(this.env, inputs, utils_1.generateBuildOptions(this.metadata, options));
            }
            yield this.runCordova(utils_1.filterArgumentsForCordova(this.metadata, inputs, options), { showExecution: true });
        });
    }
};
BuildCommand = tslib_1.__decorate([
    command_1.CommandMetadata({
        name: 'build',
        type: 'project',
        description: 'Build (prepare + compile) an Ionic project for a given platform',
        longDescription: `
Like running ${chalk_1.default.green('cordova build')} directly, but also builds web assets and provides friendly checks.

To pass additional options to the Cordova CLI, use the ${chalk_1.default.green('--')} separator after the Ionic CLI arguments. For example, for verbose log output from Cordova during an iOS build, one would use ${chalk_1.default.green('ionic cordova build ios -- -d')}. See additional examples below.
  `,
        exampleCommands: [
            'ios',
            'ios --prod --release',
            'ios --device --prod --release -- --developmentTeam="ABCD" --codeSignIdentity="iPhone Developer" --packageType="app-store"',
            'android',
            'android --prod --release -- -- --keystore=filename.keystore --alias=myalias',
            'android --prod --release -- -- --minSdkVersion=21',
            'android --prod --release -- -- --versionCode=55',
            'android --prod --release -- -- --gradleArg=-PcdvBuildMultipleApks=true',
        ],
        inputs: [
            {
                name: 'platform',
                description: `The platform to build (${['android', 'ios'].map(v => chalk_1.default.green(v)).join(', ')})`,
            }
        ],
        options: [
            // Build Options
            {
                name: 'build',
                description: 'Do not invoke an Ionic build',
                type: Boolean,
                default: true,
            },
            ...app_scripts_1.APP_SCRIPTS_OPTIONS,
            // Cordova Options
            {
                name: 'debug',
                description: 'Create a Cordova debug build',
                type: Boolean,
                intents: [utils_1.CORDOVA_INTENT],
            },
            {
                name: 'release',
                description: 'Create a Cordova release build',
                type: Boolean,
                intents: [utils_1.CORDOVA_INTENT],
            },
            {
                name: 'device',
                description: 'Create a Cordova build for a device',
                type: Boolean,
                intents: [utils_1.CORDOVA_INTENT],
            },
            {
                name: 'emulator',
                description: 'Create a Cordova build for an emulator',
                type: Boolean,
                intents: [utils_1.CORDOVA_INTENT],
            },
            {
                name: 'buildConfig',
                description: 'Use the specified Cordova build configuration',
                intents: [utils_1.CORDOVA_INTENT],
                advanced: true,
            },
        ]
    })
], BuildCommand);
exports.BuildCommand = BuildCommand;
