// ionic/dist/commands/cordova/platform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk_1 = require("chalk");
const lib_1 = require("@ionic/cli-framework/lib");
const command_1 = require("@ionic/cli-utils/lib/command");
const errors_1 = require("@ionic/cli-utils/lib/errors");
const base_1 = require("./base");
let PlatformCommand = class PlatformCommand extends base_1.CordovaCommand {
    preRun(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.preRunChecks();
            if (options['r'] || options['noresources']) {
                options['resources'] = false;
            }
            inputs[0] = (typeof inputs[0] === 'undefined') ? 'ls' : inputs[0];
            inputs[0] = (inputs[0] === 'rm') ? 'remove' : inputs[0];
            inputs[0] = (inputs[0] === 'list') ? 'ls' : inputs[0];
            lib_1.validate(inputs[0], 'action', [lib_1.contains(['add', 'remove', 'update', 'ls', 'check', 'save'], {})]);
            // If the action is list, check, or save, then just end here.
            if (['ls', 'check', 'save'].includes(inputs[0])) {
                yield this.runCordova(['platform', inputs[0]], { showExecution: true });
                throw new errors_1.FatalException('', 0);
            }
            if (!inputs[1]) {
                const platform = yield this.env.prompt({
                    type: 'input',
                    name: 'platform',
                    message: `What platform would you like to ${inputs[0]} (${['android', 'ios'].map(v => chalk_1.default.green(v)).join(', ')}):`,
                });
                inputs[1] = platform.trim();
            }
            lib_1.validate(inputs[1], 'platform', [lib_1.validators.required]);
        });
    }
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { getPlatforms } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/cordova/project'); });
            const { filterArgumentsForCordova } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/cordova/utils'); });
            let [action, platformName] = inputs;
            const platforms = yield getPlatforms(this.env.project.directory);
            if (action === 'add' && platforms.includes(platformName)) {
                this.env.log.info(`Platform ${platformName} already exists.`);
                return;
            }
            const optionList = filterArgumentsForCordova(this.metadata, inputs, options);
            if ((action === 'add' || action === 'remove') && !optionList.includes('--save')) {
                optionList.push('--save');
            }
            if (action === 'add') {
                const { installPlatform } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/cordova/project'); });
                yield installPlatform(this.env, platformName);
            }
            else {
                const response = yield this.runCordova(optionList, { showExecution: true });
                this.env.log.msg(response);
            }
            const isLoggedIn = yield this.env.session.isLoggedIn();
            if (isLoggedIn && action === 'add' && options['resources'] && ['ios', 'android'].includes(platformName)) {
                yield this.env.runCommand(['cordova', 'resources', platformName, '--force']);
            }
            this.env.tasks.end();
        });
    }
};
PlatformCommand = tslib_1.__decorate([
    command_1.CommandMetadata({
        name: 'platform',
        type: 'project',
        description: 'Manage Cordova platform targets',
        longDescription: `
Like running ${chalk_1.default.green('cordova platform')} directly, but adds default Ionic icons and splash screen resources (during ${chalk_1.default.green('add')}) and provides friendly checks.
  `,
        exampleCommands: ['', 'add ios', 'add android', 'rm ios'],
        inputs: [
            {
                name: 'action',
                description: `${chalk_1.default.green('add')}, ${chalk_1.default.green('remove')}, or ${chalk_1.default.green('update')} a platform; ${chalk_1.default.green('ls')}, ${chalk_1.default.green('check')}, or ${chalk_1.default.green('save')} all project platforms`,
            },
            {
                name: 'platform',
                description: `The platform that you would like to add (${['android', 'ios'].map(v => chalk_1.default.green(v)).join(', ')})`,
            }
        ],
        options: [
            {
                name: 'resources',
                description: `Do not pregenerate icons and splash screen resources (corresponds to ${chalk_1.default.green('add')})`,
                type: Boolean,
                default: true,
            },
        ]
    })
], PlatformCommand);
exports.PlatformCommand = PlatformCommand;
