// ionic/dist/commands/build.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk_1 = require("chalk");
const command_1 = require("@ionic/cli-utils/lib/command");
const app_scripts_1 = require("@ionic/cli-utils/lib/ionic-angular/app-scripts");
let BuildCommand = class BuildCommand extends command_1.Command {
    preRun(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (inputs.length > 0 && ['android', 'ios', 'wp8', 'windows', 'browser'].includes(inputs[0])) {
                this.env.log.warn(`${chalk_1.default.green('ionic build')} is for building web assets and takes no arguments. See ${chalk_1.default.green('ionic build --help')}.\n` +
                    `Ignoring argument ${chalk_1.default.green(inputs[0])}. Perhaps you meant ${chalk_1.default.green('ionic cordova build ' + inputs[0])}?\n`);
                inputs.splice(0);
            }
        });
    }
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { build } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/commands/build'); });
            yield build(this.env, inputs, options);
        });
    }
};
BuildCommand = tslib_1.__decorate([
    command_1.CommandMetadata({
        name: 'build',
        type: 'project',
        description: 'Build web assets and prepare your app for any platform targets',
        longDescription: `
${chalk_1.default.green('ionic build')} will perform an Ionic build, which compiles web assets and prepares them for deployment. For Ionic/Cordova apps, the CLI will run ${chalk_1.default.green('cordova prepare')}, which copies the built web assets into the Cordova platforms that you've installed. For full details, see ${chalk_1.default.green('ionic cordova prepare --help')}.
  `,
        exampleCommands: [
            '',
            '--prod',
        ],
        options: [
            ...app_scripts_1.APP_SCRIPTS_OPTIONS,
        ],
    })
], BuildCommand);
exports.BuildCommand = BuildCommand;
