// ionic/dist/commands/serve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk_1 = require("chalk");
const command_1 = require("@ionic/cli-utils/lib/command");
const serve_1 = require("@ionic/cli-utils/lib/serve");
let ServeCommand = class ServeCommand extends command_1.Command {
    preRun(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (options['nolivereload']) {
                this.env.log.warn(`The ${chalk_1.default.green('--nolivereload')} option has been deprecated. Please use ${chalk_1.default.green('--no-livereload')}.`);
                options['livereload'] = false;
            }
            if (options['nobrowser']) {
                this.env.log.warn(`The ${chalk_1.default.green('--nobrowser')} option has been deprecated. Please use ${chalk_1.default.green('--no-open')}.`);
                options['open'] = false;
            }
            if (options['b']) {
                options['open'] = false;
            }
            if (options['noproxy']) {
                this.env.log.warn(`The ${chalk_1.default.green('--noproxy')} option has been deprecated. Please use ${chalk_1.default.green('--no-proxy')}.`);
                options['proxy'] = false;
            }
            if (options['x']) {
                options['proxy'] = false;
            }
        });
    }
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { serve } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/commands/serve'); });
            yield serve(this.env, inputs, options);
        });
    }
};
ServeCommand = tslib_1.__decorate([
    command_1.CommandMetadata({
        name: 'serve',
        type: 'project',
        description: 'Start a local dev server for app dev/testing',
        longDescription: `
Easily spin up a development server which launches in your browser. It watches for changes in your source files and automatically reloads with the updated build.

By default, ${chalk_1.default.green('ionic serve')} boots up a development server on all network interfaces and prints the external address(es) on which your app is being served. It also broadcasts your app to the Ionic DevApp on your network. To disable the DevApp and bind to ${chalk_1.default.green('localhost')}, use ${chalk_1.default.green('--local')}.

Try the ${chalk_1.default.green('--lab')} option to see multiple platforms at once.
  `,
        exampleCommands: ['-c', '--lab -c'],
        options: [
            {
                name: 'consolelogs',
                description: 'Print app console logs to Ionic CLI',
                type: Boolean,
                aliases: ['c'],
            },
            {
                name: 'serverlogs',
                description: 'Print dev server logs to Ionic CLI',
                type: Boolean,
                aliases: ['s'],
                visible: false,
            },
            {
                name: 'address',
                description: 'Use specific address for the dev server',
                default: serve_1.BIND_ALL_ADDRESS,
                advanced: true,
            },
            {
                name: 'port',
                description: 'Use specific port for HTTP',
                default: String(serve_1.DEFAULT_SERVER_PORT),
                aliases: ['p'],
                advanced: true,
            },
            {
                name: 'livereload',
                description: 'Spin up server to live-reload www files',
                type: Boolean,
                default: true,
                visible: false,
            },
            {
                name: 'livereload-port',
                description: 'Use specific port for live-reload',
                default: String(serve_1.DEFAULT_LIVERELOAD_PORT),
                aliases: ['r'],
                advanced: true,
            },
            {
                name: 'dev-logger-port',
                description: 'Use specific port for dev server communication',
                default: String(serve_1.DEFAULT_DEV_LOGGER_PORT),
                advanced: true,
            },
            {
                name: 'devapp',
                description: 'Do not publish DevApp service',
                type: Boolean,
                default: true,
                advanced: true,
            },
            {
                name: 'open',
                description: 'Do not open a browser window',
                type: Boolean,
                default: true,
            },
            {
                name: 'local',
                description: 'Disable external network usage',
                type: Boolean,
            },
            {
                name: 'proxy',
                description: 'Do not add proxies',
                type: Boolean,
                default: true,
                advanced: true,
            },
            {
                name: 'browser',
                description: `Specifies the browser to use (${serve_1.BROWSERS.map(b => chalk_1.default.green(b)).join(', ')})`,
                aliases: ['w'],
                advanced: true,
            },
            {
                name: 'browseroption',
                description: `Specifies a path to open to (${chalk_1.default.green('/#/tab/dash')})`,
                aliases: ['o'],
                advanced: true,
            },
            {
                name: 'lab',
                description: 'Test your apps on multiple platform types in the browser',
                type: Boolean,
                aliases: ['l'],
            },
            {
                name: 'platform',
                description: `Start serve with a specific platform (${['android', 'ios'].map(t => chalk_1.default.green(t)).join(', ')})`,
                aliases: ['t'],
            },
            {
                name: 'auth',
                description: 'HTTP Basic Auth password to secure the server on your local network',
                type: String,
                visible: false,
            },
        ],
    })
], ServeCommand);
exports.ServeCommand = ServeCommand;
