// ionic/dist/commands/config/set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk_1 = require("chalk");
const command_1 = require("@ionic/cli-utils/lib/command");
let ConfigSetCommand = class ConfigSetCommand extends command_1.Command {
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { set } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/commands/config/set'); });
            yield set(this.env, inputs, options);
        });
    }
};
ConfigSetCommand = tslib_1.__decorate([
    command_1.CommandMetadata({
        name: 'set',
        type: 'global',
        description: 'Set config values',
        longDescription: `
By default, this command sets JSON properties in your project's ${chalk_1.default.bold('ionic.config.json')} file.

For ${chalk_1.default.green('--global')} config, the CLI sets properties in the global CLI config file (${chalk_1.default.bold('~/.ionic/config.json')}).

For nested properties, separate nest levels with dots. For example, the property name ${chalk_1.default.green('user.email')} will look in the ${chalk_1.default.bold('user')} object (a root-level field in the global CLI config file) for the ${chalk_1.default.bold('email')} field.

${chalk_1.default.green('ionic config set')} will attempt to coerce ${chalk_1.default.green('value')} into a suitable JSON type. If it is JSON-parsable, such as ${chalk_1.default.green('true')} or ${chalk_1.default.green('[]')}, it takes the parsed result. Otherwise, the value is interpreted as a string. For stricter input, use ${chalk_1.default.green('--json')}, which will error with non-JSON values.

By default, if ${chalk_1.default.green('property')} exists and is an object or an array, the value is not overwritten. To disable this check and always overwrite the property, use ${chalk_1.default.green('--force')}.
  `,
        inputs: [
            {
                name: 'property',
                description: 'The property name you wish to set',
                required: true,
            },
            {
                name: 'value',
                description: 'The new value of the given property',
                required: true,
            },
        ],
        options: [
            {
                name: 'global',
                description: 'Use global CLI config',
                type: Boolean,
                aliases: ['g'],
            },
            {
                name: 'json',
                description: `Always interpret ${chalk_1.default.green('value')} as JSON`,
                type: Boolean,
            },
            {
                name: 'force',
                description: 'Always overwrite existing values',
                type: Boolean,
                advanced: true,
            },
        ],
        exampleCommands: ['name newAppName', 'name "\\"newAppName\\"" --json', 'watchPatterns "[]" --force', '-g yarn true'],
    })
], ConfigSetCommand);
exports.ConfigSetCommand = ConfigSetCommand;
