// ionic/dist/commands/monitoring/syncmaps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = require("path");
const chalk_1 = require("chalk");
const cli_utils_1 = require("@ionic/cli-utils");
const guards_1 = require("@ionic/cli-utils/guards");
const command_1 = require("@ionic/cli-utils/lib/command");
const fs_1 = require("@ionic/cli-framework/utils/fs");
let MonitoringSyncSourcemapsCommand = class MonitoringSyncSourcemapsCommand extends command_1.Command {
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const token = yield this.env.session.getUserToken();
            const appId = yield this.env.project.loadAppId();
            const [snapshotId] = inputs;
            const { ConfigXml } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/cordova/config'); });
            const conf = yield ConfigXml.load(this.env.project.directory);
            const cordovaInfo = conf.getProjectInfo();
            const appVersion = cordovaInfo.version;
            const commitHash = yield this.env.shell.run('git', ['rev-parse', 'HEAD'], { cwd: this.env.project.directory });
            const sourcemapsDir = path.join(this.env.project.directory, '.sourcemaps');
            let sourcemapsExist = yield fs_1.pathExists(sourcemapsDir);
            if (!sourcemapsExist) {
                this.env.log.info('No sourcemaps found, doing build...');
                yield this.doProdBuild();
                sourcemapsExist = yield fs_1.pathExists(sourcemapsDir);
                if (!sourcemapsExist) {
                    this.env.log.error('Unable to sync sourcemaps. Make sure you have @ionic/app-scripts version 2.1.4 or greater.');
                    return;
                }
            }
            else {
                const doNewBuild = yield this.env.prompt({
                    type: 'confirm',
                    name: 'isProd',
                    message: 'Do build before syncing?'
                });
                doNewBuild && (yield this.doProdBuild());
            }
            this.env.log.info(`Syncing SourceMaps for app version ${chalk_1.default.green(appVersion)} of ${chalk_1.default.green(cordovaInfo.id)} (snapshot: ${snapshotId})- App ID ${appId}`);
            fs_1.readDir(sourcemapsDir).then(files => {
                const maps = files.filter(f => f.indexOf('.js.map') >= 0);
                Promise.all(maps.map(f => this.syncSourcemap(path.join(sourcemapsDir, f), snapshotId, appVersion, commitHash, appId, token)));
            });
        });
    }
    syncSourcemap(file, snapshotId, appVersion, commitHash, appId, token) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { createFatalAPIFormat } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/http'); });
            const { req } = yield this.env.client.make('POST', `/monitoring/${appId}/sourcemaps`);
            req
                .set('Authorization', `Bearer ${token}`)
                .send({
                name: path.basename(file),
                version: appVersion,
                commit: commitHash,
                snapshot_id: snapshotId
            });
            try {
                this.env.log.info(`Syncing ${chalk_1.default.green(file)}`);
                const res = yield this.env.client.do(req);
                if (res.meta.status !== 201) {
                    throw createFatalAPIFormat(req, res);
                }
                return this.uploadSourcemap(res, file);
            }
            catch (e) {
                if (guards_1.isSuperAgentError(e)) {
                    this.env.log.error(`Unable to sync map ${file}: ` + e.message);
                    if (e.response.status === 401) {
                        this.env.log.error('Try logging out and back in again.');
                    }
                    this.env.tasks.fail();
                }
                else {
                    throw e;
                }
            }
        });
    }
    uploadSourcemap(res, file) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { createRequest } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/http'); });
            const r = res;
            const fileData = yield fs_1.fsReadFile(file, { encoding: 'utf8' });
            const sourcemapPost = r.data.sourcemap_post;
            this.env.log.info('Doing this thing');
            this.env.log.info(yield this.env.config.getAPIUrl());
            let { req } = yield createRequest(this.env.config, 'post', sourcemapPost.url);
            req = req
                .buffer()
                .field(sourcemapPost.fields)
                .field('file', fileData)
                .on('progress', (event) => {
            })
                .end((err, res) => {
                if (err) {
                    this.env.log.error('Unable to upload sourcemap');
                    this.env.log.error(err);
                    return Promise.reject(err);
                }
                if (res.status !== 204) {
                    return Promise.reject(new Error(`Unexpected status code from AWS: ${res.status}`));
                }
                this.env.log.ok('Uploaded sourcemap');
                this.env.log.info('See the Error Monitoring docs for usage information and next steps: http://ionicframework.com/docs/pro/monitoring/');
                Promise.resolve();
            });
        });
    }
    doProdBuild() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const isProd = yield this.env.prompt({
                type: 'confirm',
                name: 'isProd',
                message: 'Do full prod build?'
            });
            const { build } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/commands/build'); });
            return yield build(this.env, [], { _: [], prod: isProd });
        });
    }
};
MonitoringSyncSourcemapsCommand = tslib_1.__decorate([
    command_1.CommandMetadata({
        name: 'syncmaps',
        type: 'project',
        backends: [cli_utils_1.BACKEND_PRO],
        description: 'Sync Source Maps to Ionic Pro Error Monitoring service',
        inputs: [
            {
                name: 'snapshot_id',
                description: `An Ionic Pro snapshot id to associate these sourcemaps with.`,
                required: false
            }
        ],
    })
], MonitoringSyncSourcemapsCommand);
exports.MonitoringSyncSourcemapsCommand = MonitoringSyncSourcemapsCommand;
